lib/puppet/ssl/host.rb

```diff
@@ -192,14 +195,108 @@ def generate_certificate_request(options = {})
+    end
+
+    file_path = cert_location(cert_name)
+    if Puppet::FileSystem.exist?(file_path)
+      return Puppet::SSL::Certificate.from_s(Puppet::FileSystem.read(file_path))
+    end
+
+    if Puppet::SSL::Host.ca_location == :remote
+      # Cert not found on disk, and we have a remote ca,
+      # so attempt to download it and save it to file_path
+      download_and_save_cert(cert_name, file_path)
+    end
+
+    # Couldn't find or fetch a cert
+    return nil
+  end
+
+  # Logic for detecting the cert's location on disk, based on `ca_location`
+  # and whether or not we are looking for the CA cert.
+  #
+  # @param [String] cert_name the name of the cert to find
+  # @return [String] filesystem location of the requested cert. Used both
+  # for loading and for saving.
+  def cert_location(cert_name)
+    if Puppet::SSL::Host.ca_location == :only
+      if cert_name == 'ca'
+        file_path = Puppet.settings[:cacert]
+      else
+        file_path = File.join(Puppet.settings[:signeddir], "#{name}.pem")
+      end
+    else
+      if cert_name == 'ca'
+        file_path = Puppet.settings[:localcacert]
+      else
+        file_path = File.join(Puppet.settings[:certdir], "#{name}.pem")
+      end
+    end
+    return file_path
+  end
+  private :cert_location
+
+  # Downloads the cert from the CA and saves it at the specified
+  # location on disk if the download was successful.
+  #
+  # @param [String] cert_name name of the cert to download
+  # @param [String] file_path location to save the cert
+  # return Puppet::SSL::Certificate if saved successfully, nil otherwise
+  def download_and_save_cert(cert_name, file_path)
+    response = Puppet::Rest::Routes.get_certificate(cert_name)
+    return save_cert(response, file_path)
+  end
+  private :download_and_save_cert
+
+  # Saves the body of `response` to disk at `file_path`,
+  # if the response contains one or more valid certs.
+  #
+  # @param [String] cert_name name of the cert to download
+  # @param [String] file_path location to save the cert
+  # @return Puppet::SSL::Certificate if successful, nil otherwise
+  def save_cert(response, file_path)
+    if response.ok?
+      _, body = Puppet::Rest::ResponseHandler.parse_response(response)
+      # Convert to cert object to ensure valid response
+      cert = Puppet::SSL::Certificate.from_s(body)
+      # this should be an atomic write
+      Puppet::FileSystem.open(file_path, nil, 'w:UTF-8') do |f|
+        f.puts(body)
+      end
+      return cert
+    else
+      return nil
+    end
+  end
+  private :save_cert
+
   def certificate
     unless @certificate
       generate_key unless key
 
       # get the CA cert first, since it's required for the normal cert
       # to be of any use.
-      return nil unless Certificate.indirection.find("ca", :fail_on_404 => true) unless ca?
-      return nil unless @certificate = Certificate.indirection.find(name)
+      if !ca?
+        return nil unless get_cert('ca')
```

Comment:
This should error if the CA cert can't be found. And we should verify that callers can handle the error (given the `fail_on_404` in the original, that seems likely.